README.md
<p align="center">
  <a href="https://zenaton.com" target="_blank">
    <img src="https://user-images.githubusercontent.com/36400935/58254828-e5176880-7d6b-11e9-9094-3f46d91faeee.png" target="_blank" />
  </a><br>
  Easy Asynchronous Jobs Manager for Developers <br>
  <a href="https://zenaton.com/documentation/php/getting-started/" target="_blank">
    <strong> Explore the docs » </strong>
  </a> <br>
  <a href="https://zenaton.com" target="_blank"> Website </a>
    ·
  <a href="https://github.com/zenaton/examples-php" target="_blank"> Examples in PHP </a>
    ·
  <a href="https://app.zenaton.com/tutorial/php" target="_blank"> Tutorial in PHP </a>
</p>
<p align="center">
  <a href="https://packagist.org/packages/zenaton/zenaton-php" rel="nofollow" target="_blank"><img src="https://img.shields.io/packagist/v/zenaton/zenaton-php.svg" alt="Packagist" style="max-width:100%;"></a>
  <a href="https://circleci.com/gh/zenaton/zenaton-php/tree/master" rel="nofollow" target="_blank"><img src="https://img.shields.io/circleci/project/github/zenaton/zenaton-php/master.svg" alt="CircleCI" style="max-width:100%;"></a>
  <a href="/LICENSE" target="_blank"><img src="https://img.shields.io/badge/License-Apache%202.0-blue.svg" alt="License" style="max-width:100%;"></a>
</p>

# Zenaton library for PHP

[Zenaton](https://zenaton.com) helps developers to easily run, monitor and orchestrate background jobs on your workers without managing a queuing system. In addition to this, a monitoring dashboard shows you in real-time tasks executions and helps you to handle errors.

The Zenaton library for PHP lets you code and launch tasks using Zenaton platform, as well as write workflows as code. You can sign up for an account on [Zenaton](https://zenaton.com) and go through the [tutorial in PHP](https://app.zenaton.com/tutorial/php).

## PHP Documentation

You can find all details on [Zenaton's website](https://zenaton.com/documentation/php/getting-started#introduction).

<details>
  <summary><strong>Table of contents</strong></summary>

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Getting started](#getting-started)
  - [Installation](#installation)
    - [Install the Zenaton Agent](#install-the-zenaton-agent)
    - [Install the library](#install-the-library)
    - [Framework integration](#framework-integration)
  - [Quick start](#quick-start)
    - [Client Initialization](#client-initialization)
    - [Executing a background job](#executing-a-background-job)
  - [Orchestrating background jobs](#orchestrating-background-jobs)
    - [Using workflows](#using-workflows)
- [Getting help](#getting-help)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

</details>

## Getting started

### Installation

#### Install the Zenaton Agent

To install the Zenaton agent, run the following command:

```sh
curl https://install.zenaton.com/ | sh
```

Then, you need your agent to listen to your application.
To do this, you need your **Application ID** and **API Token**.
You can find both on [your Zenaton account](https://app.zenaton.com/api).

```sh
zenaton listen --app_id=YourApplicationId --api_token=YourApiToken --app_env=YourApplicationEnv
```

#### Install the library

To add the latest version of the library to your project, run the following command:

```sh
composer require zenaton/zenaton-php
```

#### Framework integration

If you are using **Laravel** or **Symfony**, please refer to our dedicated documentation to get started:

- [Getting started with Laravel](https://zenaton.com/documentation/php/agents#laravel)
- [Getting started with Symfony](https://zenaton.com/documentation/php/agents#symfony)

### Quick start

#### Client Initialization

To start, you need to initialize the client. To do this, you need your **Application ID** and **API Token**.
You can find both on [your Zenaton account](https://app.zenaton.com/api).

Then, initialize your Zenaton client:

```php
Zenaton\Client::init('YourApplicationId', 'YourApiToken', 'YourApplicationEnv');
```

#### Executing a background job

A background job in Zenaton is a class implementing the `Zenaton\Interfaces\TaskInterface` interface.

Let's start by implementing a first task printing something, and returning a value:

```php
use Zenaton\Interfaces\TaskInterface;
use Zenaton\Traits\Zenatonable;

class HelloWorldTask implements TaskInterface
{
    public function handle()
    {
        echo "Hello World\n";

        return mt_rand(0, 1);
    }
}
```

Now, when you want to run this task as a background job, you need to do the following:

```php
(new HelloWorldTask())->dispatch();
```

That's all you need to get started. With this, you can run many background jobs.
However, the real power of Zenaton is to be able to orchestrate these jobs. The next section will introduce you to job orchestration.

### Orchestrating background jobs

Job orchestration is what allows you to write complex business workflows in a simple way.
You can execute jobs sequentially, in parallel, conditionally based on the result of a previous job,
and you can even use loops to repeat some tasks.

We wrote about some use-cases of job orchestration, you can take a look at [these articles](https://medium.com/zenaton/tagged/php)
to see how people use job orchestration.

#### Using workflows

A workflow in Zenaton is a class implementing the `Zenaton\Interfaces\WorkflowInterface` interface.

We will implement a very simple workflow:

First, it will execute the `HelloWorld` task.
The result of the first task will be used to make a condition using an `if` statement.
When the returned value will be greater than `0`, we will execute a second task named `FinalTask`.
Otherwise, we won't do anything else.

One important thing to remember is that your workflow implementation **must** be idempotent.
You can read more about that in our [documentation](https://zenaton.com/documentation/php/workflow-basics/#implementation).

The implementation looks like this:

```php
use Zenaton\Interfaces\WorkflowInterface;
use Zenaton\Traits\Zenatonable;

class MyFirstWorkflow implements WorkflowInterface
{
    use Zenatonable;

    public function handle()
    {
        $n = (new HelloWorldTask())->execute();
        if ($n > 0) {
            (new FinalTask())->execute();
        }
    }
}
```

Now that your workflow is implemented, you can execute it by calling the `dispatch` method:

```php
(new MyFirstWorkflow())->dispatch();
```

If you really want to run this example, you will need to implement the `FinalTask` task.

There are many more features usable in workflows in order to get the orchestration done right. You can learn more
in our [documentation](https://zenaton.com/documentation/php/workflow-basics/#implementation).

## Getting help

**Need help**? Feel free to contact us by chat on [Zenaton](https://zenaton.com/).

**Found a bug?** You can open a [GitHub issue](https://github.com/zenaton/zenaton-php/issues).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)

## [Unreleased]

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## [0.6.1] - 2019-10-31

### Fixed

* Fixed a possible error when API does not return errors in a correct format.

## [0.6.0] - 2019-10-02

### Changed

* Dispatch of tasks and workflows are now done using the API instead of a local agent.
* Pause, Resume and Kill workflows are now done using the API instead of a local agent.
* Send event to workflow is now done using the API instead of a local agent.
* Find workflow is now done using the API instead of a local agent.

## [0.5.1] - 2019-10-02

### Added

* Added `custom_id` argument for workflow schedule.

## [0.5.0] - 2019-08-26

### Changed

* Method `Zenaton\Traits\Zenatonable::schedule()` now has a return type of `void` instead of
  `Zenaton\Model\Scheduling\Schedule`.

## [0.4.0] - 2019-08-23

### Added

* Added an `intent_id` property when dispatching workflows and tasks, sending events to workflows, and
  pausing/resuming/killing workflows.
* Added a `::getContext()` method in `Zenatonable` trait that is able to retrieve the runtime context
  of the workflow or task currently being executed.

### Changed

* Changed `Properties` class to never serialize the `context` property from tasks and workflows.

## [0.3.4] - 2019-06-05

### Added

* Added `event_data` property when sending event.

### Removed

* Removed dependency `vlucas/phpdotenv`.

## [0.3.3] - 2019-04-16

### Fixed

* Fixed an error when serializing objects having a private `::__clone()` method.

## [0.3.2] - 2019-04-15

### Changed

* Serialization of resources now results in the integer `0` instead of throwing
  an exception in order to be consistent with the behavior of the PHP serialize
  function.

## [0.3.1] - 2019-03-29

### Fixed

* Fixed an issue with the serializer not serializing private properties defined in parent classes of the object to serialize. (#33)
* Fixed an issue with the serializer adding extra properties to objects implementing `Traversable` interface.

## [0.3.0] - 2019-03-25

### Added

* Added this changelog.
* Added unit tests.
* Added PHP CS Fixer config file and run it on all the codebase.

### Changed

* `Client::findWorkflow()` now returns `null` instead of throwing an exception when the workflow cannot be found.
* Using `Zenatonable::dispatch()` on a task outside of a workflow now dispatches the task execution instead of executing it synchronously.
* Changed `Client::ZENATON_API_URL` constant value to `https://api.zenaton.com/v1`.

### Fixed

* `Client` now properly encode parameters sent in query strings.
* `Zenaton\Workflow\Version` class is now aliased to avoid namespace shadowing bug in PHP 5.6. See <https://bugs.php.net/bug.php?id=66862>.
* `Wait::monday()` when already a monday will now wait for the next monday instead of expiring immediately, except when
  a specific time is set with `::at()` and is not already past. This behavior is also implemented in related methods
  `::tuesday()`, `::wednesday()`, `::thursday()`, `::friday()`, `::saturday()` and `::sunday()`, and also in the
  `::dayOfMonth()` method.

[Unreleased]: https://github.com/zenaton/zenaton-php/compare/0.6.1...HEAD
[0.6.1]: https://github.com/zenaton/zenaton-php/compare/0.6.0...0.6.1
[0.6.0]: https://github.com/zenaton/zenaton-php/compare/0.5.1...0.6.0
[0.5.1]: https://github.com/zenaton/zenaton-php/compare/0.5.0...0.5.1
[0.5.0]: https://github.com/zenaton/zenaton-php/compare/0.4.0...0.5.0
[0.4.0]: https://github.com/zenaton/zenaton-php/compare/0.3.4...0.4.0
[0.3.4]: https://github.com/zenaton/zenaton-php/compare/0.3.3...0.3.4
[0.3.3]: https://github.com/zenaton/zenaton-php/compare/0.3.2...0.3.3
[0.3.2]: https://github.com/zenaton/zenaton-php/compare/0.3.1...0.3.2
[0.3.1]: https://github.com/zenaton/zenaton-php/compare/0.3.0...0.3.1
[0.3.0]: https://github.com/zenaton/zenaton-php/compare/0.2.4...0.3.0


src/Zenaton/v1/Client.php
<?php

namespace Zenaton;

use Ramsey\Uuid\UuidFactory;
use Ramsey\Uuid\UuidFactoryInterface;
use Zenaton\Api\GraphQL\Mutations;
use Zenaton\Api\GraphQL\Queries;
use Zenaton\Exceptions\ApiException;
use Zenaton\Exceptions\InvalidArgumentException;
use Zenaton\Interfaces\EventInterface;
use Zenaton\Interfaces\TaskInterface;
use Zenaton\Interfaces\WorkflowInterface;
use Zenaton\Services\Http;
use Zenaton\Services\Properties;
use Zenaton\Services\Serializer;
use Zenaton\Traits\SingletonTrait;
use Zenaton\Workflows\Version as VersionedWorkflow; // Alias is required because of a bug in PHP 5.6 namespace shadowing. See https://bugs.php.net/bug.php?id=66862.

class Client
{
    use SingletonTrait;

    const ZENATON_API_URL = 'https://api.zenaton.com/v1';
    const ZENATON_GATEWAY_URL = 'https://gateway.zenaton.com/api';
    const ZENATON_WORKER_URL = 'http://localhost';
    const DEFAULT_WORKER_PORT = 4001;
    const WORKER_API_VERSION = 'v_newton';

    const MAX_ID_SIZE = 256;

    const APP_ENV = 'app_env';
    const APP_ID = 'app_id';
    const API_TOKEN = 'api_token';

    const PROG = 'PHP';

    /** @var null|string */
    protected $appId;
    /** @var null|string */
    protected $apiToken;
    /** @var null|string */
    protected $appEnv;
    /** @var Api\GraphQL\Client */
    protected $graphqlClient;
    /** @var Serializer */
    protected $serializer;
    /** @var Properties */
    protected $properties;
    /** @var UuidFactoryInterface */
    protected $uuidFactory;

    public static function init($appId, $apiToken, $appEnv)
    {
        Client::getInstance()
            ->setAppId($appId)
            ->setApiToken($apiToken)
            ->setAppEnv($appEnv)
        ;
    }

    public function construct()
    {
        $this->graphqlClient = new Api\GraphQL\Client(new Http(), $this->getGatewayUrl(), []);
        $this->serializer = new Serializer();
        $this->properties = new Properties();
        $this->uuidFactory = new UuidFactory();
    }

    /**
     * Set App Id.
     *
     * @param string $appId
     *
     * @return $this
     */
    public function setAppId($appId)
    {
        $this->appId = $appId;

        return $this;
    }

    /**
     * Set Api Token.
     *
     * @param string $apiToken
     *
     * @return $this
     */
    public function setApiToken($apiToken)
    {
        $this->apiToken = $apiToken;

        return $this;
    }

    /**
     * Set App environment.
     *
     * @param string $appEnv
     *
     * @return $this
     */
    public function setAppEnv($appEnv)
    {
        $this->appEnv = $appEnv;

        return $this;
    }

    /**
     * Return the worker url to use for a given resource.
     *
     * @param string       $resource
     * @param array|string $params
     *
     * @return string
     *
     * @internal Used by the Zenaton agent. Should not be called by user code.
     */
    public function getWorkerUrl($resource = '', $params = [])
    {
        if (is_array($params)) {
            return $this->getWorkerUrlV2($resource, $params);
        }

        $url = (getenv('ZENATON_WORKER_URL') ?: self::ZENATON_WORKER_URL)
            .':'.(getenv('ZENATON_WORKER_PORT') ?: self::DEFAULT_WORKER_PORT)
            .'/api/'.self::WORKER_API_VERSION
            .'/'.$resource.'?';

        return $this->addAppEnv($url, $params);
    }

    /**
     * Return the website url to use for a given resource.
     *
     * @param string       $resource
     * @param array|string $params
     *
     * @return string
     *
     * @internal Used by the Zenaton agent. Should not be called by user code.
     */
    public function getWebsiteUrl($resource = '', $params = [])
    {
        if (is_array($params)) {
            return $this->getWebsiteUrlV2($resource, $params);
        }

        $url = $this->getApiUrl()
            .'/'.$resource.'?'
            .self::API_TOKEN.'='.$this->apiToken.'&';

        return $this->addAppEnv($url, $params);
    }

    /**
     * Start a single task.
     *
     * @throws ApiException if the API returns some errors
     */
    public function startTask(TaskInterface $task)
    {
        $this->sendGatewayRequestAndThrowOnErrors(function () use ($task) {
            return $this->graphqlClient->request(
                Mutations::DISPATCH_TASK,
                [
                    'input' => [
                        'environmentName' => $this->appEnv,
                        'intentId' => $this->uuidFactory->uuid4()->toString(),
                        'maxProcessingTime' => method_exists($task, 'getMaxProcessingTime') ? $task->getMaxProcessingTime() : null,
                        'programmingLanguage' => self::PROG,
                        'name' => \get_class($task),
                        'data' => $this->serializer->encode($this->properties->getPropertiesFromObject($task)),
                    ],
                ],
                [
                    'App-Id' => $this->appId,
                    'Api-Token' => $this->apiToken,
                ]
            );
        });
    }

    /**
     * Start a workflow instance.
     *
     * @param WorkflowInterface $flow Workflow to start
     *
     * @throws ApiException             if the API returns some errors
     * @throws InvalidArgumentException if custom id is invalid
     */
    public function startWorkflow(WorkflowInterface $flow)
    {
        $canonical = null;
        if ($flow instanceof VersionedWorkflow) {
            $canonical = \get_class($flow);
            // replace $flow by true current implementation
            $flow = $flow->getCurrentImplementation();
        }
        $customId = $this->getCustomIdFromWorkflow($flow);

        $this->sendGatewayRequestAndThrowOnErrors(function () use ($customId, $canonical, $flow) {
            return $this->graphqlClient->request(
                Mutations::DISPATCH_WORKFLOW,
                [
                    'input' => [
                        'customId' => $customId,
                        'environmentName' => $this->appEnv,
                        'intentId' => $this->uuidFactory->uuid4()->toString(),
                        'programmingLanguage' => self::PROG,
                        'canonicalName' => $canonical,
                        'data' => $this->serializer->encode($this->properties->getPropertiesFromObject($flow)),
                        'name' => \get_class($flow),
                    ],
                ],
                [
                    'App-Id' => $this->appId,
                    'Api-Token' => $this->apiToken,
                ]
            );
        });
    }

    /**
     * Kill a workflow instance.
     *
     * @param string $workflowName Workflow class name
     * @param string $customId     Provided custom id
     *
     * @throws ApiException if the API returns some errors
     */
    public function killWorkflow($workflowName, $customId)
    {
        $this->sendGatewayRequestAndThrowOnErrors(function () use ($customId, $workflowName) {
            return $this->graphqlClient->request(
                Mutations::KILL_WORKFLOW,
                [
                    'input' => [
                        'customId' => $customId,
                        'environmentName' => $this->appEnv,
                        'intentId' => $this->uuidFactory->uuid4()->toString(),
                        'programmingLanguage' => self::PROG,
                        'name' => $workflowName,
                    ],
                ],
                [
                    'App-Id' => $this->appId,
                    'Api-Token' => $this->apiToken,
                ]
            );
        });
    }

    /**
     * Pause a workflow instance.
     *
     * @param string $workflowName Workflow class name
     * @param string $customId     Provided custom id
     *
     * @throws ApiException if the API returns some errors
     */
    public function pauseWorkflow($workflowName, $customId)
    {
        $this->sendGatewayRequestAndThrowOnErrors(function () use ($customId, $workflowName) {
            return $this->graphqlClient->request(
                Mutations::PAUSE_WORKFLOW,
                [
                    'input' => [
                        'customId' => $customId,
                        'environmentName' => $this->appEnv,
                        'intentId' => $this->uuidFactory->uuid4()->toString(),
                        'programmingLanguage' => self::PROG,
                        'name' => $workflowName,
                    ],
                ],
                [
                    'App-Id' => $this->appId,
                    'Api-Token' => $this->apiToken,
                ]
            );
        });
    }

    /**
     * Resume a workflow instance.
     *
     * @param string $workflowName Workflow class name
     * @param string $customId     Provided custom id
     *
     * @throws ApiException if the API returns some errors
     */
    public function resumeWorkflow($workflowName, $customId)
    {
        $this->sendGatewayRequestAndThrowOnErrors(function () use ($customId, $workflowName) {
            return $this->graphqlClient->request(
                Mutations::RESUME_WORKFLOW,
                [
                    'input' => [
                        'customId' => $customId,
                        'environmentName' => $this->appEnv,
                        'intentId' => $this->uuidFactory->uuid4()->toString(),
                        'programmingLanguage' => self::PROG,
                        'name' => $workflowName,
                    ],
                ],
                [
                    'App-Id' => $this->appId,
                    'Api-Token' => $this->apiToken,
                ]
            );
        });
    }

    /**
     * Schedule a task instance.
     *
     * @param string $cron
     *
     * @throws ApiException
     */
    public function scheduleTask(TaskInterface $task, $cron)
    {
        $this->sendGatewayRequestAndThrowOnErrors(function () use ($task, $cron) {
            return $this->graphqlClient->request(
                Mutations::CREATE_TASK_SCHEDULE,
                [
                    'input' => [
                        'environmentName' => $this->appEnv,
                        'cron' => $cron,
                        'intentId' => $this->uuidFactory->uuid4()->toString(),
                        'programmingLanguage' => static::PROG,
                        'properties' => $this->serializer->encode($this->properties->getPropertiesFromObject($task)),
                        'taskName' => \get_class($task),
                    ],
                ],
                [
                    'App-Id' => $this->appId,
                    'Api-Token' => $this->apiToken,
                ]
            );
        });
    }

    /**
     * Schedule a workflow instance.
     *
     * @param string $cron
     *
     * @throws ApiException
     * @throws InvalidArgumentException
     */
    public function scheduleWorkflow(WorkflowInterface $workflow, $cron)
    {
        $canonicalName = $name = \get_class($workflow);
        if ($workflow instanceof VersionedWorkflow) {
            $workflow = $workflow->getCurrentImplementation();
            $name = \get_class($workflow);
        }
        $customId = $this->getCustomIdFromWorkflow($workflow);

        $this->sendGatewayRequestAndThrowOnErrors(function () use ($workflow, $canonicalName, $cron, $name, $customId) {
            return $this->graphqlClient->request(
                Mutations::CREATE_WORKFLOW_SCHEDULE,
                [
                    'input' => [
                        'environmentName' => $this->appEnv,
                        'cron' => $cron,
                        'canonicalName' => $canonicalName,
                        'intentId' => $this->uuidFactory->uuid4()->toString(),
                        'customId' => $customId,
                        'programmingLanguage' => static::PROG,
                        'properties' => $this->serializer->encode($this->properties->getPropertiesFromObject($workflow)),
                        'workflowName' => $name,
                    ],
                ],
                [
                    'App-Id' => $this->appId,
                    'Api-Token' => $this->apiToken,
                ]
            );
        });
    }

    /**
     * Find a workflow instance.
     *
     * @param string $workflowName Workflow class name
     * @param string $customId     Provided custom id
     *
     * @throws ApiException if the API returns some errors
     *
     * @return null|WorkflowInterface
     */
    public function findWorkflow($workflowName, $customId)
    {
        $response = $this->graphqlClient->request(
            Queries::WORKFLOW,
            [
                'customId' => $customId,
                'environmentName' => $this->appEnv,
                'programmingLanguage' => self::PROG,
                'workflowName' => $workflowName,
            ],
            [
                'App-Id' => $this->appId,
                'Api-Token' => $this->apiToken,
            ]
        );

        if (isset($response['errors'])) {
            // If there is a NOT_FOUND error, we return null
            $notFoundErrors = \array_filter($response['errors'], static function ($error) {
                return isset($error['type']) && 'NOT_FOUND' === $error['type'];
            });
            if (\count($notFoundErrors) > 0) {
                return null;
            }

            // Otherwise, we generate an exception from the error list
            throw ApiException::fromErrorList($response['errors']);
        }

        return $this->properties->getObjectFromNameAndProperties($response['data']['findWorkflow']['name'], $this->serializer->decode($response['data']['findWorkflow']['properties']));
    }

    /**
     * Send an event to a workflow instance.
     *
     * @param string         $workflowName Workflow class name
     * @param string         $customId     Provided custom id
     * @param EventInterface $event        Event to send
     *
     * @throws ApiException if the API returns some errors
     */
    public function sendEvent($workflowName, $customId, EventInterface $event)
    {
        $this->sendGatewayRequestAndThrowOnErrors(function () use ($workflowName, $customId, $event) {
            return $this->graphqlClient->request(
                Mutations::SEND_EVENT,
                [
                    'input' => [
                        'customId' => $customId,
                        'environmentName' => $this->appEnv,
                        'name' => get_class($event),
                        'input' => $this->serializer->encode($this->properties->getPropertiesFromObject($event)),
                        'data' => $this->serializer->encode($event),
                        'intentId' => $this->uuidFactory->uuid4()->toString(),
                        'programmingLanguage' => self::PROG,
                        'workflowName' => $workflowName,
                    ],
                ],
                [
                    'App-Id' => $this->appId,
                    'Api-Token' => $this->apiToken,
                ]
            );
        });
    }

    /**
     * @param string $url
     * @param string $params
     *
     * @return string
     *
     * @deprecated 0.3.0 Does not properly encode url parameters.
     */
    protected function addAppEnv($url, $params = '')
    {
        static $triggeredDeprecation = false;
        if (!$triggeredDeprecation) {
            trigger_error('You are running a Zenaton agent version <= 0.4.5 which is using deprecated code. Please consider upgrading your agent.', E_USER_DEPRECATED);
        }
        $triggeredDeprecation = true;

        // when called from worker, APP_ENV and APP_ID is not defined
        return $url
            .($this->appEnv ? self::APP_ENV.'='.$this->appEnv.'&' : '')
            .($this->appId ? self::APP_ID.'='.$this->appId.'&' : '')
            .($params ? $params.'&' : '');
    }

    /**
     * @param string $url
     *
     * @return string
     */
    protected function addQueryParams($url, array $params = [])
    {
        // When called from worker, APP_ENV and APP_ID are not defined
        if ($this->appId) {
            $params[static::APP_ID] = $this->appId;
        }
        if ($this->appEnv) {
            $params[static::APP_ENV] = $this->appEnv;
        }

        return sprintf('%s?%s', $url, http_build_query($params));
    }

    /**
     * @param string $resource
     *
     * @return string
     */
    private function getWorkerUrlV2($resource = '', array $params = [])
    {
        $url = sprintf(
            '%s:%s/api/%s/%s',
            getenv('ZENATON_WORKER_URL') ?: static::ZENATON_WORKER_URL,
            getenv('ZENATON_WORKER_PORT') ?: static::DEFAULT_WORKER_PORT,
            static::WORKER_API_VERSION,
            $resource
        );

        return $this->addQueryParams($url, $params);
    }

    /**
     * @param string $ressources
     *
     * @return string
     */
    private function getWebsiteUrlV2($ressources = '', array $params = [])
    {
        $url = sprintf(
            '%s/%s',
            $this->getApiUrl(),
            $ressources
        );

        $params[static::API_TOKEN] = $this->apiToken;

        return $this->addQueryParams($url, $params);
    }

    /**
     * Returns the Zenaton API url to send requests to.
     *
     * This will check the ZENATON_API_URL environment variable first if it is defined. Otherwise, it will use the
     * default production url.
     *
     * @return string
     */
    private function getApiUrl()
    {
        return \getenv('ZENATON_API_URL') ?: self::ZENATON_API_URL;
    }

    /**
     * Returns the Zenaton Gateway url to send requests to.
     *
     * This will check the ZENATON_GATEWAY_URL environment variable first if it is defined. Otherwise, it will use the
     * default production url.
     *
     * @return string
     */
    private function getGatewayUrl()
    {
        return \getenv('ZENATON_GATEWAY_URL') ?: self::ZENATON_GATEWAY_URL;
    }

    /**
     * Sends a GraphQL request and throw an exception if there are errors returned by the API.
     *
     * @throws ApiException
     *
     * @return array
     */
    private function sendGatewayRequestAndThrowOnErrors(\Closure $closure)
    {
        $response = $closure();

        if (isset($response['errors'])) {
            throw ApiException::fromErrorList($response['errors']);
        }

        return $response;
    }

    /**
     * Returns the custom id of a workflow.
     *
     * @throws InvalidArgumentException
     *
     * @return null|string
     */
    private function getCustomIdFromWorkflow(WorkflowInterface $workflow)
    {
        if (!\method_exists($workflow, 'getId')) {
            return null;
        }

        $customId = $workflow->getId();
        if (!\is_string($customId) && !\is_int($customId)) {
            throw new InvalidArgumentException('Provided Id must be a string or an integer');
        }
        // convert it to a string
        $customId = (string) $customId;
        // should be not more than 256 bytes;
        if (\strlen($customId) > self::MAX_ID_SIZE) {
            throw new InvalidArgumentException('Provided Id must not exceed '.self::MAX_ID_SIZE.' bytes');
        }

        return $customId;
    }
}
